azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -966,18 +980,27 @@ def _run_op(self, op):
     @pipeline_thread.runs_on_pipeline_thread
     def _handle_pipeline_event(self, event):
         if isinstance(event, pipeline_events_base.DisconnectedEvent):
-            if self.pipeline_root.connected:
-                logger.info(
-                    "{}({}): State is {}.  Triggering reconnect timer".format(
-                        self.name, event.name, self.state
-                    )
+            logger.debug(
+                "{}({}): State is {} Connected is {}.".format(
+                    self.name, event.name, self.state, self.pipeline_root.connected
                 )
+            )
+
+            if self.pipeline_root.connected and self.state == ReconnectState.LOGICALLY_CONNECTED:
+                # when we get disconnected, we try to immediatly reconnect.  If that fails,
+                # then we wait a while before retrying.
+                #
+                # For now, we use immediate_reconnect_delay to mean "close enough to immediate"
+                # because it saves us from refactoring this code when we know that another change
+                # is coming.  Specifically, we know that this stage is broken in cases where the
+                # user calls disconnect().  In that case, this stage doesn't currently know that
+                # it shouldn't reconnect automatically.
                 self.state = ReconnectState.WAITING_TO_RECONNECT
-                self._start_reconnect_timer()
+                self._start_reconnect_timer(self.immediate_reconnect_delay)
+
             else:
```

Comment:
>else [](start = 12, length = 4)

for an user disconnect the code path will turn up here right ? where the reconnect timer will not fire ?

Reply:
corrrect

---
In reply to: [437156650](https://github.com/Azure/azure-iot-sdk-python/pull/576#discussion_r437156650) [](ancestors = 437156650)